R/clusterSystems.R

```diff
@@ -0,0 +1,146 @@
+#'
+#' @author \href{https://orcid.org/0000-0001-5724-2252}{Rachel Silverstein} (aut)
+#'
+#' @seealso For examples of distance functions that can be used to create custom distance matrices, see \code{\link{expr_dist}} \code{\link{jaccard_dist}} \code{\link{tf_dist}}.
+#'
+#' @examples
+#' systemDB_name <- "SysDB"
+#' myDB <- fetchData(systemDB_name)
+#' rootSysIDs <- SyDBgetRootSysIDs(myDB)
+#' sys_names <- names(rootSysIDs)
+#' systems <- SyDBgetSysSymbols(myDB, sys_names)
+#'
+#' # Cluster all of the systems in the database according to
+#' # the sum of transcription factor distance and Jaccard network distance:
+#' clusterSystems(systems,
+#'                distances = c("transcription_factor", "network_jaccard"),
+#'                combineMatrices = 'sum')
+#'
+#' # Cluster all of the systems using expression profile distance
+#' # but format it like a custom distance function
+#' GEO <- fetchData("GEOprofiles")
+#' clusterSystems(systems,
+#'                distances = NULL,
+#'                customDistanceFn = list(expr_dist),
+#'                dataSources = list(GEO),
+#'                combineMatrices = NULL)
+#'
+#'
+#'
+#' @import cluster
+#' @import VennDiagram
+#' @import igraph
+#'
+#' @export
+
+
+clusterSystems <- function(systems,
+                           distances,
+                           customDistanceFn = NULL,
+                           dataSources = NULL,
+                           combineMatrices,
+                           printVennDiagrams = TRUE) {
+
+
+  # make a single vector containing all of the genes from all of the systems together
+  all_genes <- unique(unlist(systems))
+
+  distanceMatrices <- make_distance_matrices_list(distances, all_genes)
+
+  # compute any distance matrices using custom similarity functions and data sources
+  if (!is.null(customDistanceFn)) {
+    for (i in seq_along(customDistanceFn)) {
+      data <- dataSources[[i]]
+      fn <- customDistanceFn[[i]]
+      matrix <- make_matrix(all_genes, fn, data)
+      matrix <- list(matrix)
+      names(matrix) <- paste(c("custom_function_", i), collapse = "")
+      distanceMatrices <- append(distanceMatrices, matrix)
+    }
+  }
+
+  # Combine the distance matrices if there is more than one
+  combinedMatrix <- combine_distance_matrices(mode = combineMatrices, distanceMatrices)
+
+
+  # CLUSTER TIME :-)
+
+  # use the number of systems as the number of clusters for PAM clustering
+  clust_obj <- cluster::pam(combinedMatrix, k = length(systems), diss = TRUE)
+
+  clusters <- clust_obj$clustering
+
+  # get the best cluster - system pairings
+  matches <- get_best_matches(systems, clusters)
+  best_matches <- matches[[1]]
+  jaccard_indexes <- matches[[2]]
+
+  # Calcuulate P values of getting Jaccard index this high or higher for each
+  # system - cluster pair
+  p_values <- jaccard_index_p_values(systems, jaccard_indexes, best_matches, clusters, trials = 1000)
+
+  result <- list(clusters, best_matches, jaccard_indexes, p_values)
```

Comment:
You can combine these two statements as `result <- list(a = "A", b = "B")`... or even
```
result <- list()
result$a <- "A"
result$b <- "B"
```
but defining the list first, then naming it has a greater risk that name and contents get disconnected.